python/ray/ml/utils/tensorflow_utils.py

```diff
@@ -0,0 +1,86 @@
+import numpy as np
+import pandas as pd
+import tensorflow as tf
+
+from ray.data.extensions.tensor_extension import TensorDtype
+
+
+def convert_pandas_to_tf_tensor(df: pd.DataFrame) -> tf.Tensor:
+    """Convert a pandas dataframe to a TensorFlow tensor.
+
+    This function works in two steps:
+    1. Convert each dataframe column to a tensor.
+    2. Concatenate the resulting tensors along the last axis.
+
+    Arguments:
+        df: The dataframe to convert to a TensorFlow tensor. Columns must have
+            a numeric dtype, ``TensorDtype``, or object dtype. If a column has
+            an object dtype, the column must contain ``ndarray`` objects.
+
+    Returns:
+        A tensor of data type ``float32`` constructed from the dataframe.
+
+    Raises:
+        ValueError: if a column has an invalid dtype.
+        ValueError: if the columns can't be combined into a single tensor.
+
+    Examples:
+        >>> import pandas as pd
+        >>> from ray.ml.utils.tensorflow_utils improt convert_pandas_to_tf_tensor
+        >>>
+        >>> df = pd.DataFrame({"X1": [1, 2, 3], "X2": [4, 5, 6]})
+        >>> convert_pandas_to_tf_tensor(df[["X1"]]).shape
+        TensorShape([3])
+        >>> convert_pandas_to_tf_tensor(df[["X1", "X2"]]).shape
+        TensorShape([3, 2])
+
+        >>> from ray.data.extensions import TensorArray
+        >>>
+        >>> df = pd.DataFrame({"image": TensorArray(np.zeros((4, 3, 32, 32)))})
+        >>> convert_pandas_to_tf_tensor(df).shape
+        TensorShape([4, 3, 32, 32])
+    """
+    def is_valid_dtype(series) -> bool:
+        if pd.api.types.is_numeric_dtype(series) or isinstance(series, TensorDtype):
+            return True
+
+        is_ndarray = series.map(lambda obj: isinstance(obj, np.ndarray))
+        return all(is_ndarray)
```

Comment:
Left a comment on your example PR that the idiomatic way to support ndarray elements in a column should be using our tensor column extension type, although I think that generically supporting opaque ndarray objects in Pandas columns is still desirable until we can ensure that our extension type is always used.

Also, IMO we should make sure that this check short-circuits so we don't need to do a full batch scan here:
```suggestion
        return all(isinstance(obj, np.ndarray) for _, obj in series.iteritems())
```